llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp

```diff
@@ -0,0 +1,116 @@
+    std::vector<llvm::Instruction *> InstrDeleteList;
+    llvm::LLVMContext &CTX = Func.getContext();
+    for (llvm::BasicBlock &Block : Func) {
+      for (llvm::Instruction &Instr : Block) {
+        if (llvm::CallInst *CallInstruction =
+                llvm::dyn_cast<llvm::CallInst>(&Instr)) {
+          llvm::Function *Called = CallInstruction->getCalledFunction();
+          if (Called == &Func || !Called->getReturnType()->isVoidTy() ||
+              Called->getFunctionType()->getNumParams() > 0) {
+            continue;
+          }
+          std::map<llvm::BasicBlock *, llvm::BasicBlock *> BlockMap;
+          llvm::ValueToValueMapTy VMap;
+          llvm::BasicBlock *SplittedBlock = splitBlockBefore(&Block, &Instr);
+          llvm::BasicBlock *NextBlockPlace = SplittedBlock->getNextNode();
+          for (llvm::BasicBlock &ToInlineBasicBlock : *Called) {
+            llvm::BasicBlock *TmpBlock =
+                llvm::BasicBlock::Create(CTX, "", &Func, NextBlockPlace);
+            BlockMap[&ToInlineBasicBlock] = TmpBlock;
+            NextBlockPlace = TmpBlock->getNextNode();
+          }
+          for (llvm::BasicBlock &ToInlineBasicBlock : *Called) {
+            for (llvm::Instruction &ToInlineInstr : ToInlineBasicBlock) {
+              llvm::BasicBlock *CurrentBlock = BlockMap[&ToInlineBasicBlock];
+              if (llvm::isa<llvm::ReturnInst>(&ToInlineInstr)) {
+                llvm::BranchInst::Create(&Block)->insertInto(
+                    CurrentBlock, CurrentBlock->end());
+              } else {
+                llvm::Instruction *NewInstr = ToInlineInstr.clone();
+                if (llvm::isa<llvm::BranchInst>(*NewInstr)) {
+                  unsigned OpNum = NewInstr->getNumOperands();
+                  for (unsigned Ind = 0; Ind < OpNum; Ind++) {
+                    llvm::Value *CurOperand = NewInstr->getOperand(Ind);
+                    if (llvm::isa<llvm::BasicBlock>(*CurOperand)) {
+                      NewInstr->setOperand(
+                          Ind, llvm::dyn_cast<llvm::Value>(
+                                   BlockMap[llvm::dyn_cast<llvm::BasicBlock>(
+                                       CurOperand)]));
+                    }
+                  }
+                }
+                NewInstr->insertInto(CurrentBlock, CurrentBlock->end());
+                llvm::RemapInstruction(NewInstr, VMap,
+                                       llvm::RF_NoModuleLevelChanges |
+                                           llvm::RF_IgnoreMissingLocals);
+                VMap[&ToInlineInstr] = NewInstr;
+              }
+            }
+          }
+          SplittedBlock->getTerminator()->setOperand(
+              0,
+              llvm::dyn_cast<llvm::Value>(BlockMap[&Called->getEntryBlock()]));
+          InstrDeleteList.push_back(&Instr);
+        }
+      }
+    }
+    for (llvm::Instruction *Instr : InstrDeleteList) {
+      Instr->eraseFromParent();
+    }
+    return llvm::PreservedAnalyses::all();
```

Comment:
Are you sure that results of all analyses are not staled after such a code transformation?

Reply:
> Are you sure that results of all analyses are not staled after such a code transformation?

It passes circular call (tested in test8.ll) and I haven't came up with the situation where this pass fails, so I think yes. For each run it doesn't change any other functions other than the one that's presented on this run.

Reply:
Analysis passes results are used by not only this pass, but in case of pipelines of passes they may be used by any further pass after yours one. If you don't reset analysis results, other passes may start to use all function structure information and their result may even corrupt valid behavior of compiled program.

Reply:
> Analysis passes results are used by not only this pass, but in case of pipelines of passes they may be used by any further pass after yours one. If you don't reset analysis results, other passes may start to use all function structure information and their result may even corrupt valid behavior of compiled program.

I don't really know. It generates the correct IR code, the naming is correct, the dependency is also correct (names do not cross in a single function). I don't see a reason why any other pass might fail. Is it because I wrap inline function between 2 non-conditional branch instructions so that inline function will always be in a separate block?

Reply:
This is not about current pas, this is about cached results of other analysis passes https://llvm.org/docs/Passes.html.
Just search for preserved analyses info.